### PR TITLE
actions: coverage

### PR DESCRIPTION
**Pull request by oliver-sanders**
closes: #3598
follow on from: #3899

Reinstates code coverage for Cylc.

It does:

* Capture all `cylc` commands.
* Write the coverage files in your Cylc working directory.
* Rsync coverage files back for code coverage on GH actions.
* Plug some minor coverage holes that were quick to write tests for.

It does not:

* Capture coverage for direct access to the `cylc` library from Python (i.e. not via the `cylc` CLI).
    * For this you can use standard tools directly (`coverage`, `pytest --cov`, etc).
* Capture anything after the daemonisation (i.e. when `--no-detach` is *not* set).
    * Can be done in the future though there is little to gain from doing so.
* Rsync coverage files back for code coverage locally.
    * Only does this on GH actions.

For assurance that a `cylc` command is being captured and covered set `CYLC_COVERAGE=2`, you will get a `.coverage_commands_captured` file which lists each command for which coverage was successfully recorded.

Initial results suggest we have [~89.5% coverage](https://codecov.io/gh/oliver-sanders/cylc-flow/tree/3d5803ff7a34cc02e44cf4849a9b7fe64df6ddff/cylc/flow) with obvious holes in the coverage around remote functionality, for now...

...Remote results in, 91.22%, not bad.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (it is tests).
- [x] No change log entry required (invisible to users).
- [x] No documentation update required (documented in the code).
- [x] No dependency changes.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
This should get that test run on the right platform:

```
diff --git a/tests/functional/remote/04-symlink-dirs.t b/tests/functional/remote/04-symlink-dirs.t
index 1e343d709..bb706cbf3 100644
--- a/tests/functional/remote/04-symlink-dirs.t
+++ b/tests/functional/remote/04-symlink-dirs.t
@@ -17,8 +17,8 @@
 #-------------------------------------------------------------------------------
 # Checks configured symlinks are created for run, work, share, share/cycle, log 
 # # directories on localhost and the remote platform.
-. "$(dirname "$0")/test_header"
 export REQUIRE_PLATFORM='loc:remote comms:tcp'
+. "$(dirname "$0")/test_header"
 
 if [[ -z ${TMPDIR:-} || -z ${USER:-} || $TMPDIR/$USER == "$HOME" ]]; then
```

**Comment by oliver-sanders** [comment]
Deconflicted

File: cylc/flow/scripts/cylc.py [hunk 1]
```diff
@@ -362,74 +363,170 @@ def cli_version(long=False):
...
+    $ cd /cylc/working/directory
+    $ coverage combine
+    $ coverage report
+
+    For remote tasks the coverage files will be written to the cylc
+    working directory on the remote so you will have to scp them back
+    to your local working directory before running coverage combine:
+
+    $ cd /cylc/working/directory
+    $ ssh remote-host cd /cylc/remote/working/directory && covrage combine
```
Comment (MetRonnie) [inline cylc/flow/scripts/cylc.py:392]: Typo
```suggestion
    $ ssh remote-host cd /cylc/remote/working/directory && coverage combine
```

File: cylc/flow/scripts/cylc.py [hunk 2]
```diff
@@ -362,74 +363,170 @@ def cli_version(long=False):
...
+        yield
+        return
+
+    # import here to avoid unnecessary imports when not running coverage
+    import cylc.flow
+    import coverage
+    from pathlib import Path
+
+    # the cylc working directory
+    cylc_wc = Path(cylc.flow.__file__).parent.parent.parent
```
Comment (MetRonnie) [inline cylc/flow/scripts/cylc.py:423]:
```suggestion
    cylc_wc = Path(cylc.flow.__file__).parents[2]
```
https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.parents

**Comment by MetRonnie** [comment]
`tests/unit/main_loop/auto_restart.py → tests/unit/main_loop/test_auto_restart` has lost its extension

**Review by MetRonnie — CHANGES_REQUESTED** [review]
`tests/unit/main_loop/auto_restart.py → tests/unit/main_loop/test_auto_restart` has lost its extension. Other suggestions optional

File: cylc/flow/scripts/cylc.py (hunk 1 above)
Comment (oliver-sanders) [reply cylc/flow/scripts/cylc.py:392]: Well spotted.

File: cylc/flow/scripts/cylc.py (hunk 2 above)
Comment (oliver-sanders) [reply cylc/flow/scripts/cylc.py:423]: Nice

**Review by MetRonnie — APPROVED** [review]

File: cylc/flow/scripts/cylc.py [hunk 3]
```diff
@@ -362,74 +363,170 @@ def cli_version(long=False):
...
+                    else:
+                        dd_index = 0
+                    cmd_args.insert(dd_index, "--")
+
+            if help_:
+                execute_cmd(command, "--help")
+            else:
+                if version:
+                    cmd_args.append("--version")
+                execute_cmd(command, *cmd_args)
```
Comment (kinow) [inline cylc/flow/scripts/cylc.py:532]: :clap: great workaround !

**Review by kinow — APPROVED** [review]

**Pull request merged**